### PR TITLE
[RISCV] Pack build_vectors into largest available element type

### PR DESCRIPTION
**Pull request by preames**
Our worst case build_vector lowering is a serial chain of vslide1down.vx operations which creates a serial dependency chain through a relatively high latency operation.  We can instead pack together elements into ELEN sized chunks, and move them from integer to scalar in a single operation.

This reduces the length of the serial chain on the vector side, and costs at most three scalar instructions per element.  This is a win for all cores when the sum of the latencies of the scalar instructions is less than the vslide1down.vx being replaced, and is particularly profitable for out-of-order cores which can overlap the scalar computation.

This patch is restricted to configurations with zba and zbb.  Without both, the zero extend might require two instructions which would bring the total scalar instructions per element to 4.  zba and zba are both present in the rva22u64 baseline which is looking to be quite common for hardware in practice; we could extend this to systems without bitmanip with a bit of extra effort.

### CODE REVIEW

**Comment by preames** [comment]
Quick note for reviewers, please pay close attention to byte order issues in this.  I frequently get that wrong, and while this is inspired by a downstream patch I rewrote this basically from scratch.  Definitely room for error.

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll
Comment (topperc) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:1270]: Should we test RVA22U32?

Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:1270]: Not sure it adds any useful coverage, but happy to add it if desired.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3943]: I think you're right, this condition is just dead.  Will remove.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3920]: Do we allow configurations where ELEN > XLEN?  I couldn't find a clear answer in the specification as to whether that was a valid point in the configuration cross product.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3920]: Zve64* on RV32 allows that

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:1270]: I asked specifically because it allows ELEN=64 with XLen=32.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3920]: Do we ever end up with a non power of 2 VT during lowering? If not we could move `NumElts % 2 != 0` to an assert

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll
Comment (lukel97) [inline llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:2416]: ptr %p looks to be unused in these tests

**Review by lukel97 — APPROVED** [review]
LGTM with nits

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:3920]: 1 element vectors?

File: llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll
Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:1270]: We don't appear to have support in tree for rva22u32 (as an mattr alias).  I went digging through profile docs, and I can't find any mention of the rva22u32 profile variants; rva22 appears to only define 64 bit versions.

Do you have a particular arch string for a configuration which exercises the case you're concerned about here?

Comment (preames) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:1270]: From your other comment, it seems like plain rv32 + v + zba + zbb might be enough here?  Can you confirm?

Comment (topperc) [reply llvm/test/CodeGen/RISCV/rvv/fixed-vectors-int-buildvec.ll:1270]: I hadn't realized that that the RVA* profiles are 64-bit only. rv32 + v + zba + zbb is enough.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: I added tests for 1 element vectors, just to be sure we had coverage here.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp]: Added, and rebased.  Turned out to be a good thing, as the configuration exposed a crash (which is now fixed).

**Comment by preames** [comment]
@topperc Have I addressed all of your concerns?  This has an LGTM, but you pointed out a substantial problem after that, so I don't want to land without confirmation.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**
